File: src/main/java/net/azisaba/townia/data/TowniaPlayer.java
package net.azisaba.townia.data;

import java.util.UUID;

public class TowniaPlayer {

    private final UUID uuid;
    private String name;
    private UUID townUuid;
    private TownRank rank;
    private long lastSeen;
    private String preferredLang;
    private java.util.List<String> friends = new java.util.ArrayList<>();

    public TowniaPlayer(UUID uuid, String name, UUID townUuid, TownRank rank, long lastSeen, String preferredLang) {
        this.uuid = uuid;
        this.name = name;
        this.townUuid = townUuid;
        this.rank = rank;
        this.lastSeen = lastSeen;
        this.preferredLang = preferredLang;
    }
    
    public TowniaPlayer(UUID uuid, String name, UUID townUuid, TownRank rank, long lastSeen, String preferredLang, java.util.List<String> friends) {
        this.uuid = uuid;
        this.name = name;
        this.townUuid = townUuid;
        this.rank = rank;
        this.lastSeen = lastSeen;
        this.preferredLang = preferredLang;
        if (friends != null) this.friends = friends;
    }

    public UUID getUuid() { return uuid; }
    public String getName() { return name; }
    public UUID getTownUuid() { return townUuid; }
    public TownRank getRank() { return rank; }
    public long getLastSeen() { return lastSeen; }
    public String getPreferredLang() { return preferredLang; }
    public java.util.List<String> getFriends() { return friends; }

    public boolean isInTown() { return townUuid != null; }
    public boolean isMayor() { return rank == TownRank.MAYOR; }
    public boolean isCoMayor() { return rank == TownRank.CO_MAYOR; }
    public boolean isAssistant() { return rank.isAtLeast(TownRank.ASSISTANT); }
    public boolean isAssistantOrHigher() { return rank.isAtLeast(TownRank.ASSISTANT); }
    public boolean isMayorOrHigher() { return rank.isAtLeast(TownRank.MAYOR); }

    public void setName(String name) { this.name = name; }
    public void setTownUuid(UUID townUuid) { this.townUuid = townUuid; }
    public void setRank(TownRank rank) { this.rank = rank; }
    public void setLastSeen(long lastSeen) { this.lastSeen = lastSeen; }
    public void setPreferredLang(String preferredLang) { this.preferredLang = preferredLang; }
    public void setFriends(java.util.List<String> friends) { this.friends = friends; }
}

File: src/main/java/net/azisaba/townia/data/TowniaPlayer.kt
package net.azisaba.townia.data

import java.util.*

class TowniaPlayer {
    var uuid: UUID?
    var name: String?
    var townUuid: UUID?
    var rank: TownRank
    var lastSeen: Long
    var preferredLang: String?
    var friends: MutableList<String?>? = ArrayList<String?>()

    var isToggleMap: Boolean = false
    var isToggleTownClaim: Boolean = false
    var isTogglePlotBorder: Boolean = false

    var defaultPermsFriend: String? = ""
    var defaultPermsAlly: String? = ""
    var defaultPermsOutsider: String? = ""
    var defaultPermsResident: String? = "BDSI"

    constructor(uuid: UUID?, name: String?, townUuid: UUID?, rank: TownRank, lastSeen: Long, preferredLang: String?) {
        this.uuid = uuid
        this.name = name
        this.townUuid = townUuid
        this.rank = rank
        this.lastSeen = lastSeen
        this.preferredLang = preferredLang
    }

    constructor(
        uuid: UUID?,
        name: String?,
        townUuid: UUID?,
        rank: TownRank,
        lastSeen: Long,
        preferredLang: String?,
        friends: MutableList<String?>?
    ) {
        this.uuid = uuid
        this.name = name
        this.townUuid = townUuid
        this.rank = rank
        this.lastSeen = lastSeen
        this.preferredLang = preferredLang
        if (friends != null) this.friends = friends
    }

    val isInTown: Boolean
        get() = townUuid != null
    val isMayor: Boolean
        get() = rank == TownRank.MAYOR
    val isCoMayor: Boolean
        get() = rank == TownRank.CO_MAYOR
    val isAssistant: Boolean
        get() = rank.isAtLeast(TownRank.ASSISTANT)
    val isAssistantOrHigher: Boolean
        get() = rank.isAtLeast(TownRank.ASSISTANT)
    val isMayorOrHigher: Boolean
        get() = rank.isAtLeast(TownRank.MAYOR)
}